pkg/broker/handler/processors/deliver/processor.go

```diff
@@ -147,23 +147,79 @@ func (p *Processor) Process(ctx context.Context, e *event.Event) error {
 
 // deliver delivers msg to target and sends the target's reply to the broker ingress.
 func (p *Processor) deliver(ctx context.Context, target *config.Target, broker *config.CellTenant, msg binding.Message, hops int32) error {
+	// Channels can have a reply address without a subscriber. So default the replyMessage to the
+	// original message. If there is a subscriber, then replyMessage is overwritten.
+	replyMessage := msg
+	if target.Address != "" {
```

Comment:
Great catch. Added a special case and a unit test.

Reply:
So after the latest commit, we will nack the message to the delivery pubsub in the broker topic and then reply on pubsub to redeliver it, right? The previous behavior is move to the DLQ instead. I think both way are good, but just want to make sure we understand the difference.

Reply:
This behavior should be identical to before. In both cases, an empty string address results in `deliver` returning an error. In the past, this was because we actually tried sending the event and the HTTP call failed saying `""` is a bad address https://github.com/google/knative-gcp/blob/c19b557da0480a1806b07243bcbb76dd7ee12413/pkg/broker/handler/processors/deliver/processor.go#L152-L160
Now we just generate the error directly. 

The DLQ is handled by the PubSub layer, not the delivery processor. Both before and after this PR. When we create the Subscription, we instruct PubSub to send to a DLQ after N failed deliveries.

Reply:
Sorry, my bad, I meant to say retry queue instead of DLQ. You are right both case it will be sent to the retry topic in lin 142.